internal/provider/resource_access_offer.go

```diff
@@ -0,0 +1,436 @@
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read offer %s, got error: %s", offerURL, err))
+		return
+	}
+	a.trace(fmt.Sprintf("read juju offer %q", offerURL))
+
+	// Add to state the ones in the state and set in the offer
+	users := []string{}
+	var access string
+	for _, offerUserDetail := range response.Users {
+		if offerUserDetail.UserName == "everyone@external" || offerUserDetail.UserName == "admin" {
+			continue
+		}
+		users = append(users, offerUserDetail.UserName)
+		if access != "" && access != string(offerUserDetail.Access) {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("User %s has different access %s", offerUserDetail.UserName, offerUserDetail.Access))
+			return
+		}
+		access = string(offerUserDetail.Access)
+	}
+	newStateUsers, errDiag := basetypes.NewSetValueFrom(ctx, types.StringType, users)
+	state.Users = newStateUsers
+	resp.Diagnostics.Append(errDiag...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Set the plan onto the Terraform state
+	state.ID = types.StringValue(newAccessOfferIDFrom(offerURL, access, users))
+	state.Access = types.StringValue(access)
+	state.OfferURL = types.StringValue(offerURL)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (a *accessOfferResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	// Check first if the client is configured
+	if a.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "access offer", "read")
+		return
+	}
+	var plan accessOfferResourceOffer
+
+	// Get the plan
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Retrieve information from the plan
+	offerURL := plan.OfferURL.ValueString()
+	access := plan.Access.ValueString()
+	var planUsers []string
+	if err := plan.Users.ElementsAs(ctx, &planUsers, false); err != nil {
+		resp.Diagnostics.AddError("Plan error", fmt.Sprintf("Unable to read users from the plan, got error: %s", err))
+		return
+	}
+
+	// Get user/access info from Offer
+	response, err := a.client.Offers.ReadOffer(&juju.ReadOfferInput{
+		OfferURL: offerURL,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read offer, got error: %s", err))
+		return
+	}
+	a.trace(fmt.Sprintf("read juju offer %q", offerURL))
+	offerUsers := response.Users
+
+	// Check if a user was removed and delete access (revoke read)
+	// Create map to be used in the next step for adding new users
+	offerUserNames := make(map[string]struct{})
+	for _, offerUser := range offerUsers {
+		if offerUser.UserName == "everyone@external" || offerUser.UserName == "admin" {
+			continue
+		}
+		offerUserNames[offerUser.UserName] = struct{}{}
+		if !slices.Contains(planUsers, offerUser.UserName) {
+			a.trace(fmt.Sprintf("revoke user %q for offer %q", offerUser.UserName, offerURL))
+			err := a.client.Offers.RevokeOffer(&juju.GrantRevokeOfferInput{
```

Comment:
we can't just revoke user's access without taking into account the access level - maybe this resource has nothing to do with the user: e.g. resource_access_offer specifies level `consume`, but the user has `admin` permission on the offer - as it is now, we'd revoke admin users all access from the offer

Reply:
If a user has been removed from the resource, then to remove their privileges, you'd Revoke `read`. Afaik, it's not necessary to step through the levels.

I've also suggested not allowing the `admin` user as a valid user to avoid problems there.